sqllogictest/src/parser.rs

```diff
@@ -294,6 +280,104 @@ impl<T: ColumnType> std::fmt::Display for Record<T> {
     }
 }
 
+/// Expected error message after `error` or under `----`.
+#[derive(Debug, Clone)]
+pub enum ExpectedError {
+    /// An inline regular expression after `error`.
+    ///
+    /// The actual error message that matches the regex is considered as a match.
+    Inline(Regex),
+    /// A multiline error message under `----`, ends with 2 consecutive empty lines.
+    ///
+    /// The actual error message that's exactly the same as the expected one is considered as a
+    /// match.
+    Multiline(String),
+}
+
+impl ExpectedError {
+    /// Parses an inline regex variant from tokens.
+    fn from_inline_tokens(tokens: &[&str]) -> Result<Self, ParseErrorKind> {
+        let err_str = tokens.join(" ");
+        let regex =
+            Regex::new(&err_str).map_err(|_| ParseErrorKind::InvalidErrorMessage(err_str))?;
+        Ok(Self::Inline(regex))
+    }
+
+    /// Returns whether it's an empty inline regex.
+    fn is_empty_inline(&self) -> bool {
+        match self {
+            ExpectedError::Inline(regex) => regex.as_str().is_empty(),
+            ExpectedError::Multiline(_) => false,
+        }
+    }
+
+    /// Unparses the expected message after `error`, if it's inline.
```

Comment:
nit: More strictly speaking, if so we should fmt `error` outside the function. 🤪